### PR TITLE
Upgrade React-Native website to Docusaurus v3

### PR DESCRIPTION
**Pull request by slorber**
WIP PR to upgrade to Docusaurus v3 canary

I will let you know when it's ready to review

---

Deploy preview: https://deploy-preview-3780--react-native.netlify.app/

Argos visual tests 
- Reports: https://app.argos-ci.com/slorber/rnw-visual-tests
- PR: https://github.com/slorber/rnw-visual-tests/pull/2

---

Note to myself: we can revert usage of `｜` in favor of `\|`, see change made in https://github.com/facebook/react-native-website/pull/3807#discussion_r1297318844

---

Follow-up of:
- v2.4.1 upgrade: https://github.com/facebook/react-native-website/pull/3778
- v3 docs preparation - iteration 1: https://github.com/facebook/react-native-website/pull/3807
- v3 docs preparation - iteration 2: https://github.com/facebook/react-native-website/pull/3830
- v3 docs preparation - iteration 3: https://github.com/facebook/react-native-website/pull/3843

### CODE REVIEW

**Comment by slorber** [comment]
@Simek there are docs that I had to modify so that they can compile with MDX 2.

Those are mostly isolated in this commit, only updating the last version:
[`ad86759` (#3780)](https://github.com/facebook/react-native-website/pull/3780/commits/ad86759c72f9363e50696a007c9d9ae4691cb770)
(older versions are probably also affected by the same problems)

If you use expressions like `{x: number, y: number}`, `version <5` or `<COMMIT_HASH>` in MD, you either have to escape them with `\` or find another way to create the same text.

I used escaping here, but IMHO you should rather use inline code blocks for many of these things. I also noticed that some API table types are already in a code block, and some others are not, so this is not super consistent, and MDX 2 catched that.

```md

Before, doesn't work with MDX 2
| ({nativeEvent: [PressEvent](pressevent)}) => boolean |

Escaped:
| (\{nativeEvent: [PressEvent](pressevent)}) => boolean |

Inline code block:
| `({nativeEvent: [PressEvent](pressevent)}) => boolean` |
```

My suggestion is to use inline code blocks consistently everywhere for such cases.

If you can do this right now on existing docs and merge it, that would be a nice preliminary work for this PR, and would help reduce the diff so that this PR remains easy to review.

File: plugins/remark-snackplayer/package.json
Comment (cipolleschi) [inline plugins/remark-snackplayer/package.json:27]: why this has been downgraded? 🤔

File: plugins/remark-snackplayer/package.json
Comment (cipolleschi) [inline plugins/remark-snackplayer/package.json:24]: we are removing this because we don't need it?

File: website/docusaurus.config.js
Comment (cipolleschi) [inline website/docusaurus.config.js:23]: I think that If this is not used, we can remove it.

File: website/package.json
Comment (cipolleschi) [inline website/package.json:51]: Are these version numbers correct? 🤔 
when are we going to move to 3.y.z?

**Review by cipolleschi — COMMENTED** [review]
Hi @slorber, thank you so much for this PR. I left a few comments, mainly for clarification.
The only significant piece I am not confident to review are the Snack updates: at first glance, they looks good to me, but  let's wait for another pair of eyes to double check them.

File: plugins/remark-snackplayer/package.json
Comment (slorber) [reply plugins/remark-snackplayer/package.json:24]: This is because it is used for the remark snackplayer code but it only works for MDX v1 and its usage has been removed.

CleanShot 2023-09-12 at 15 01 12@2x


MDX v1 used to be more based on "string concatenation".

With MDX v2, JSX elements have first-class MDX AST nodes that you can construct manually.

This code has been replaced by: 

CleanShot 2023-09-12 at 15 02 55@2x

File: website/package.json
Comment (slorber) [reply website/package.json:51]: @cipolleschi those are canary Docusaurus versions: https://docusaurus.io/community/canary

As soon as this PR is merged, I'll release Docusaurus v3-beta.0 and we can replace those with the first beta.

Or if you agree to merge, I can also publish v3-beta.0 first, and update this PR before merging.

My goal here is to be able to say to the community that a few major Docusaurus sites are already running on Docusaurus v3 beta (Docusaurus + Jest + React-Native), showing it is safe, and encourage adoption. 

I don't consider those canaries to be unstable myself, and could eventually release this canary as v3 as GA today. This is mostly a way to reassure the Docusaurus community and help increase adoption so that we can catch the few minor bugs that we might have missed. Similarly to when RN releases a RC and is looking for community feedback ;)

File: plugins/remark-snackplayer/package.json
Comment (slorber) [reply plugins/remark-snackplayer/package.json]: Hmm, can't remember and looks suspicious. Reverting.

File: website/docusaurus.config.js
Comment (slorber) [reply website/docusaurus.config.js]: yes 👍 removed

File: website/core/DocsRating.js
Comment (slorber) [inline website/core/DocsRating.js:13]: This fixes a little React hydration issue. The issue was already there but is now logged thanks to React 18 new hydration callback `onRecoverableError(error)` (https://react.dev/blog/2022/03/29/react-v18)

The SSR/SSG render and the first CSR render outputs should match, so things such as `typeof window !== "undefined"` etc should be avoided.

**Comment by slorber** [comment]
> Hi @slorber, thank you so much for this PR. I left a few comments, mainly for clarification. The only significant piece I am not confident to review are the Snack updates: at first glance, they looks good to me, but let's wait for another pair of eyes to double check them.

Thanks @cipolleschi

Regarding the SnackPlayer, unless you know well the differences between MDX v1/v2 it would be hard to review technically. IMHO you'd rather focus on the outcome: is the snack player still working on all the pages in the deploy preview, or do you see any regression?

File: website/package.json
Comment (cipolleschi) [reply website/package.json:51]: Thanks for the explanation. I think that they are fine as they are.

**Comment by cipolleschi** [comment]
> Regarding the SnackPlayer, unless you know well the differences between MDX v1/v2 it would be hard to review technically. IMHO you'd rather focus on the outcome: is the snack player still working on all the pages in the deploy preview, or do you see any regression?

Not really, I was just mentioning it in case someone among @cortinico @TheSavior or @Simek wants to double-check that code! :D

File: website/src/css/customTheme.scss
Comment (TheSavior) [inline website/src/css/customTheme.scss:1877]: This needs a new line at the end of the file it seems

File: website/blog/2018-03-05-AWS-app-sync.md
Comment (TheSavior) [inline website/blog/2018-03-05-AWS-app-sync.md:97]: Why does this need to be escaped?

File: website/blog/2015-03-26-react-native-bringing-modern-web-techniques-to-mobile.md
Comment (TheSavior) [inline website/blog/2015-03-26-react-native-bringing-modern-web-techniques-to-mobile.md:21]: Was this change made by a formatter or manually?

**Review by TheSavior — APPROVED** [review]
The snack changes seem fine, and the snack player seems to work in the preview

File: website/src/css/customTheme.scss
Comment (slorber) [reply website/src/css/customTheme.scss]: I added a new line but not sure why it's needed? CI is passing. Is there linter I don't see?

File: website/blog/2018-03-05-AWS-app-sync.md
Comment (slorber) [reply website/blog/2018-03-05-AWS-app-sync.md:97]: Docusaurus v3 now uses https://github.com/remarkjs/remark-directive to provide support for admonitions, and other things using syntax such as `:textDirective`, `::leafDirective` and `:::containerDirective`.

`Re:invent` is parsed as `Re` + `:invent` text directive.

This is what happens in the MDX playground if you turn the remark-directive option on:

CleanShot 2023-09-14 at 13 06 03@2x


This regression was captured by the visual regression tests:

CleanShot 2023-09-14 at 12 51 24@2x

---

Technically it should probably be possible to add some code in Docusaurus v3 so that this escaping becomes un-necessary, and the unhandled AST directive nodes get serialized back as raw text. For now it's not implemented, and not even sure who should implement that 😅. Will think about it. I hope in the meantime it's not a blocker for you to merge this PR.

**Comment by slorber** [comment]
Thanks for the review @cipolleschi @TheSavior

Let me know if I can do anything else ;)

**Review by cipolleschi — APPROVED** [review]

**Review by cipolleschi — APPROVED** [review]

File: website/src/css/customTheme.scss
Comment (TheSavior) [reply website/src/css/customTheme.scss]: I mentioned it because GitHub showed this icon presumably because there wasn't an extra new line, whereas all other files have that extra new line. 
CleanShot 2023-09-15 at 17 48 25

And yeah, it's normally enforced with a linter. See react-native's main eslint config: https://github.com/facebook/react-native/blob/2be409ff55b50563a5f123f7823fa7cdb72dbef9/packages/eslint-config-react-native/index.js#L264

**Pull request merged**
